bpf/lib/tunnel.h

```diff
@@ -83,4 +88,135 @@ tunnel_vni_to_sec_identity(__be32 vni)
+	    + sizeof(struct vxlanhdr) > data_end)
+		return false;
+
+	udp = (struct udphdr *)(data + sizeof(struct ethhdr) + l3_size);
+
+	if (udp->dest == bpf_htons(tunnel_port))
+		return true;
+
+	return false;
+}
+
+/*
+ * Returns the VNI in the native host's endian format of a xvlan encap'd packet.
+ *
+ * The caller must ensure the skb associated with these data buffers are infact
+ * a vxlan encapsulated packet before invoking this function.
+ *
+ * This can be done by calling 'tunnel_skb_is_vxlan_v4'
+ *
+ */
+static __always_inline __u32
+tunnel_vxlan_get_vni(const void *data, const void *data_end,
+		     const struct iphdr *ipv4)
+{
+	int l3_size = ipv4->ihl * 4;
+	struct vxlanhdr *hdr;
+
+	if (data + sizeof(struct ethhdr) + l3_size + sizeof(struct udphdr)
+	    + sizeof(struct vxlanhdr) > data_end)
+		return 0;
+
+	hdr = (struct vxlanhdr *)(data + sizeof(struct ethhdr) + l3_size +
+	       sizeof(struct udphdr));
+
+	return bpf_ntohl(hdr->vx_vni) >> 8;
+}
+
+/*
+ * Points 'inner' to the inner IPv4 header of a IPv4 VXLan excapsulated
+ * packet.
+ *
+ * The caller should be sure the VXLan packet is encapsulating IPv4 traffic
+ * before calling this method.
+ *
+ * Returns 'true' if 'inner' now points to a bounds-checked inner IPv4 header.
+ * Returns 'false' if an error occurred.
+ */
+static __always_inline bool
+tunnel_vxlan_get_inner_ipv4(const void *data, const void *data_end,
+			    const struct iphdr *ipv4, struct iphdr **inner) {
+	int l3_size = ipv4->ihl * 4;
+
+	if (data + sizeof(struct ethhdr) + l3_size + sizeof(struct udphdr)
+	    + sizeof(struct vxlanhdr) + sizeof(struct ethhdr) +
+	    sizeof(struct iphdr) > data_end)
+		return false;
+
+	*inner = (struct iphdr *)(data + sizeof(struct ethhdr)
+		  + l3_size + sizeof(struct udphdr) + sizeof(struct vxlanhdr)
+		  + sizeof(struct ethhdr));
+
+	return true;
+}
+
+/*
+ * Rewrites the current VNI in the VXLan header to the provided and updates
+ * the l4 checksum if necessary.
+ *
+ * The VNI is in host endian format when supplied to this function.
+ *
+ * The caller must ensure the skb associated with these data buffers are infact
+ * a vxlan encapsulated packet before invoking this function.
+ *
+ * This can be done by calling 'tunnel_skb_is_vxlan_v4'
+ */
+static __always_inline bool
+tunnel_vxlan_rewrite_vni(void *ctx, const void *data, const void *data_end,
+			 const struct iphdr *ipv4, __u32 vni)
+{
+	struct csum_offset csum = {0};
+	int l3_size = ipv4->ihl * 4;
+	int l4_off = sizeof(struct ethhdr) + l3_size;
+	struct udphdr *udp = NULL;
+	struct vxlanhdr *vx = NULL;
+	__be32 old_vni = 0;
+
+	if (data + sizeof(struct ethhdr) + l3_size + sizeof(struct udphdr)
+	    + sizeof(struct vxlanhdr) > data_end)
+		return false;
+
+	udp = (struct udphdr *)(data + sizeof(struct ethhdr) + l3_size);
+
+	vx = (struct vxlanhdr *)(data + sizeof(struct ethhdr) + l3_size
+	      + sizeof(struct udphdr));
+
+	old_vni = vx->vx_vni;
+	vx->vx_vni = bpf_htonl(vni << 8);
```

Comment:
> The userspace code comments say "store the sec_id in network byte order", but here the code is doing "host to network long" with the value from the map.

Huh, I never see a comparison to integers using ntoh like here:

```
static __always_inline bool identity_is_world_ipv6(__u32 identity)
{
#if defined ENABLE_IPV4 && defined ENABLE_IPV6
		return identity == WORLD_ID || identity == WORLD_IPV6_ID;
#else
		return identity == WORLD_ID;
#endif
}
``` 

But maybe this is because we know we are always a single byte value here?

Reply:
The usage of the function is here: 

```
	ep_info = __lookup_ip4_endpoint(inner_ipv4->saddr);
	if (!ep_info)
		return DROP_INVALID;

	/*
	 * this is a vxlan packet so ip4->daddr is the tunnel endpoint
	 */
	ret = set_ipsec_encrypt(ctx, encrypt_key, ip4->daddr, ep_info->sec_id,
				false);
	if (ret != CTX_ACT_OK)
		return ret;

	/*
	 * source mac is our current egress interface, lets copy it to dmac
	 * so redirecting to ingress side of the same interface doesn't fail.
	 */
	eth_load_saddr(ctx, &dst_mac, 0);
	eth_store_daddr(ctx, &dst_mac, 0);

	/* need to revalidate data since we just re-wrote mac addresses */
	if (!revalidate_data(ctx, &data, &data_end, &ip4))
		return DROP_INVALID;

	/* right now, the VNI of this packet is ENCRYPTED_OVERLAY_ID, we need
	 * to rewrite this VNI to the source's sec id before we transmit it
	 */
	if (!tunnel_vxlan_rewrite_vni(ctx, data, data_end, ip4,
				      ep_info->sec_id))
		return DROP_INVALID;
```

So ep_info->sec_id will be already big endian when I grab it out of the map?

Reply:
> This logic also matches the equivalent/opposite function in tunnel_vni_to_sec_identity(). 

Ah yes, very good point.